### ISSUE TITLE
Streaming compression?

### ISSUE THREAD

**Issue by mahmoud** [warner/magic-wormhole#326]
Earlier today I had to cancel a half-complete wormhole transfer to compress a large text file before resending.

Seems to me that most users (myself included) wouldn't care much if there was a transparent gzip and ungzip stream happening before encryption and after decryption, right?

**Comment by warner** [comment]
Yeah, I think that'd be just fine. The one wrinkle is that it'd be nice to know both the compressed and the uncompressed size ahead of time: the uncompressed size can be used to warn the receiving side how much space is about to be consumed if they approve the transfer, while the compressed size might let us show a more accurate progress bar.

We don't really need the compressed size, though, if we pretend that the compression ratio is roughly constant for the whole file, so using the uncompressed output stream to drive the progress bar should be good enough.

There's one other wrinkle for the upcoming "dilated wormhole" effort, in which multiple independent streams are multiplexed over the same persistent TCP connection. My plan there is to send multiple files through separate streams, kind of like how FTP used to work back in the day (one file per stream). In that world, doing an independent gzip context on each stream won't get us as much compression as our current "build a zipfile of the whole directory in RAM first" approach. And I think sharing a gzip context between independent streams is awfully difficult. So maybe, for sending directories over dilated wormholes, we should find a way to do a streaming zip+compress process, and only use a single stream.

The trickiest part is probably negotiating the ability to gzip the streams: the receiving side must advertise the ability, and the sending side must notice it. There's a place for this advertisement to go (dilated wormholes use the same mechanism), but it'll take some thought about backwards/forwards compatibility to make sure we get it right.

**Comment by DL6AKU** [comment]
I just want to throw [zstd](https://facebook.github.io/zstd/#other-languages) into the discussion. Faster and better compression ratio than gzip.
